### PR TITLE
examples/golang : Add bucket notification deletion example

### PR DESCRIPTION
**Pull request by faithuniterh**
Signed-off-by: Faith <faithuniterh@tutanota.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: examples/golang/bucket-notification-deletion/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Introduction
+This directory contains Golang code example on how to create an SNS Topic on a Ceph RGW cluster.
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/README.md:2]: this is not a topic creation example. please change

File: examples/golang/bucket-notification-deletion/README.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+# Prerequisite
+Linux machine running an RGW Ceph cluster. Preferrably started with the ``OSD=1 MON=1 MDS=0 MGR=0 RGW=1 ../src/vstart.sh --debug --new `` command.  
+Go installed on the Linux machine.  
+
+## Workflow Procedure
+1. Navigate through your file system to where the Golang example code exists on your terminal.
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/README.md:9]: to which directory exactly? there are two sub-directories here

File: examples/golang/bucket-notification-deletion/notification-deletion-example/go.mod [hunk 3]
```diff
@@ -0,0 +1,9 @@
+module notification-deletion/notification-deletion-example
+
+go 1.19
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/notification-deletion-example/go.mod:3]: why do we need go 1.19?
can we use go 1.18 (what i have on fedora36) or even go 1.17 (what i have on rhel8)?

File: examples/golang/bucket-notification-deletion/notification-deletion/go.mod [hunk 4]
```diff
@@ -0,0 +1,3 @@
+module notification-deletion/notification
+
+go 1.19
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/notification-deletion/go.mod:3]: same as https://github.com/ceph/ceph/pull/48040/files#r974061858

File: examples/golang/bucket-notification-deletion/README.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# Introduction
+This directory contains Golang code example on how to create an SNS Topic on a Ceph RGW cluster.
+
+# Prerequisite
+Linux machine running an RGW Ceph cluster. Preferrably started with the ``OSD=1 MON=1 MDS=0 MGR=0 RGW=1 ../src/vstart.sh --debug --new `` command.  
+Go installed on the Linux machine.  
+
+## Workflow Procedure
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/README.md:8]: please explain about the sub-directory structure here

File: examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+func main() {
+	bucket := flag.String("b", "", "Name of the bucket to add notification to")
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go:16]: remove not add

File: examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	topic := flag.String("t", "", "The topic onto which the notification is attached to")
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go:17]: why do we need a topic here?
we should have an optional "notificationID" indicating deletion of specific notification or all notifications of a bucket

File: examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+	flag.Parse()
+
+	if *bucket == "" {
+		fmt.Println("You must supply the name of the bucket")
+		fmt.Println("-b BUCKET")
+		return
+	}
+
+	if *topic == "" {
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go:26]: no topic is needed

File: examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+		fmt.Println("You must supply the name of the topic ARN")
+		fmt.Println("-t TOPIC ARN")
+		return
+	}
+
+	//Ceph RGW Credentials
+	access_key := "0555b35654ad1656d804"
+	secret_key := "h7GhxuBLTrlhVUyxSPUKUV8r/2EI4ngqJxD7iBdBYLhwluN30JaT3Q=="
+	token_id := ""
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go:35]: no need for token_id

File: examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go:7]: add "notification" to imports

File: examples/golang/bucket-notification-deletion/notification-deletion-example/go.mod [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+
+replace notification-deletion/notification => ../notification
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/notification-deletion-example/go.mod:5]:
```suggestion
replace notification => ../notification-deletion
```

File: examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+	svc := s3.New(sess)
+
+	input := &s3.notification.DeleteBucketNotificationRequestInput{
+		Bucket: bucket,
+	}
+
+	_, err := svc.notification.DeleteBucketNotification(input, topic)
+
+	/*input := &s3.PutBucketNotificationConfigurationInput{
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go:66]: please remove dead code

File: examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+			Region:           aws.String("default"),
+			Credentials:      credentials.NewStaticCredentials(access_key, secret_key, token_id),
+			S3ForcePathStyle: aws.Bool(true),
+			EndpointResolver: endpoints.ResolverFunc(CustResolverFn),
+		},
+	}))
+
+	svc := s3.New(sess)
+
+	input := &s3.notification.DeleteBucketNotificationRequestInput{
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go:60]: if we call the extension library "notification" we should change this to:
```golang
input := &notification.DeleteBucketNotificationRequestInput{
```

File: examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+		Bucket: bucket,
+	}
+
+	_, err := svc.notification.DeleteBucketNotification(input, topic)
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go:64]: this is our extension, should be:
```golang
err := notification.DeleteBucketNotification(scv, input, 
```

Comment (yuvalif) [reply examples/golang/bucket-notification-deletion/notification-deletion-example/notification-deletion-example.go:64]: also, no need for "topic" parameter

File: examples/golang/bucket-notification-deletion/notification-deletion/notification.go [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+import (
+	"net/http"
+
+	"github.com/aws/aws-sdk-go/aws/awsutil"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+type DeleteBucketNotificationRequestInput struct {
```
Comment (yuvalif) [inline examples/golang/bucket-notification-deletion/notification-deletion/notification.go:11]: please add an optional "notificationID" parameter

File: examples/golang/bucket-notification-deletion/notification-deletion-example/go.mod (hunk 3 above)
Comment (faithuniterh) [reply examples/golang/bucket-notification-deletion/notification-deletion-example/go.mod]: The new server I got is running go 1.19 that is why this is here.
I have changed it to go 1.18 as you advised.

File: examples/golang/bucket-notification-deletion/notification-deletion/notification.go (hunk 15 above)
Comment (faithuniterh) [reply examples/golang/bucket-notification-deletion/notification-deletion/notification.go:11]: How Do I go about adding the NotificationID here?

File: examples/golang/bucket-notification-deletion/README.md (hunk 5 above)
Comment (faithuniterh) [reply examples/golang/bucket-notification-deletion/README.md:8]: I was thinking of making the notification code as an extension that can be used in the example code.
The notification-deletion folder contains the custom Rook code written and the notification-deletion-example folder has the example code showing how a notification can be deleted.
I intended to call the code in the notification-deletion example as a custom package in the notification-deletion-example code as per the info at https://golangbot.com/go-packages/

Comment (yuvalif) [reply examples/golang/bucket-notification-deletion/README.md:8]: ok. so you have to explain in the README what to do where;
"Navigate through your file system to where the Golang example code exists on your terminal"
does not explain to which directory should we navigate

**Comment by yuvalif** [comment]
i currently get this compilation error:
```
# notification-deletion/notification-deletion-example
./notification-deletion-example.go:64:12: assignment mismatch: 2 variables but notification.DeleteBucketNotification returns 1 value
./notification-deletion-example.go:64:55: not enough arguments in call to notification.DeleteBucketNotification
        have (*s3.S3, *notification.DeleteBucketNotificationRequestInput)
        want (*s3.S3, *notification.DeleteBucketNotificationRequestInput, string)
./notification-deletion-example.go:69:53: undefined: topic

```

File: examples/golang/bucket-notification-deletion/notification-deletion/notification.go (hunk 15 above)
Comment (yuvalif) [reply examples/golang/bucket-notification-deletion/notification-deletion/notification.go:11]: add it to this type. similarly to "Bucket":
```golang
// The name of the notification to delete from the bucket.
//  
// NotificationID is an optional field. If not provided, all notifications are deleted from the bucket
NotificationID *string `location:"uri" locationName:"NotificationID" type:"string" required:"false"`
```

**Pull request closed without merging**